FirebaseCore/Internal/Sources/HeartbeatLogging/RingBuffer.swift

```diff
@@ -14,38 +14,57 @@
 
 import Foundation
 
+/// Error types for `RingBuffer` operations.
+enum RingBufferError: LocalizedError {
+  case outOfBoundsPush(index: Array.Index)
+
+  var errorDescription: String {
+    switch self {
+    case let .outOfBoundsPush(index: index):
+      return "Out-of-bounds push to ring buffer at index \(index)."
+    }
+  }
+}
+
 /// A generic circular queue structure.
 struct RingBuffer<Element>: Sequence {
   /// An array of heartbeats treated as a circular queue and intialized with a fixed capacity.
   private var circularQueue: [Element?]
   /// The current "tail" and insert point for the `circularQueue`.
-  private var tailIndex: Int = 0
+  private var tailIndex: Array.Index
 
   /// Designated initializer.
   /// - Parameter capacity: An `Int` representing the capacity.
   init(capacity: Int) {
     circularQueue = Array(repeating: nil, count: capacity)
+    tailIndex = circularQueue.startIndex
   }
 
   /// Pushes an element to the back of the buffer, returning the element (`Element?`) that was overwritten.
   /// - Parameter element: The element to push to the back of the buffer.
   /// - Returns: The element that was overwritten or `nil` if nothing was overwritten.
   /// - Complexity: O(1)
   @discardableResult
-  mutating func push(_ element: Element) -> Element? {
+  mutating func push(_ element: Element) throws -> Element? {
     guard circularQueue.count > 0 else {
       // Do not push if `circularQueue` is a fixed empty array.
       return nil
     }
 
-    defer {
-      // Increment index, wrapping around to the start if needed.
-      tailIndex += 1
-      tailIndex %= circularQueue.count
+    if !circularQueue.indices.contains(tailIndex) {
+      // We have somehow entered an invalid state (#10025).
+      throw RingBufferError.outOfBoundsPush(index: tailIndex)
```

Comment:
Might be helpful to include `endIndex` in the error as well.

Reply:
Actually, the `count`/`endIndex` is always 30 because the buffer is manually hardcoded to have a 30 element capacity. 30 was selected as the limit for the number of heartbeats stored on device so I'm not sure the number will change enough to make its inclusion in the error useful.

Reply:
Given that we can't see how `tailIndex` could ever be > 30, it seems that the more data we gather the better.

Reply:
Sorry, I'm confused by "it seems that the more data we gather the better.". So are you saying you're ok with not collecting `tailIndex` since we have no plans to change it?

Reply:
Reworked in [0af9de5](https://github.com/firebase/firebase-ios-sdk/pull/10032/commits/0af9de501a3265ee05d47d5922d9c7e67342d5cd)